src/vmm/src/devices/virtio/queue.rs

```diff
@@ -214,23 +190,70 @@ pub struct Queue {
     pub ready: bool,
 
     /// Guest physical address of the descriptor table
-    pub desc_table: GuestAddress,
+    pub desc_table_address: GuestAddress,
 
     /// Guest physical address of the available ring
-    pub avail_ring: GuestAddress,
+    pub avail_ring_address: GuestAddress,
 
     /// Guest physical address of the used ring
-    pub used_ring: GuestAddress,
-
-    pub(crate) next_avail: Wrapping<u16>,
-    pub(crate) next_used: Wrapping<u16>,
+    pub used_ring_address: GuestAddress,
+
+    /// This is a user space pointer
+    /// to the available ring, set up by the
+    /// guest. Getting access to the underling
+    /// data structure should only occur after the
+    /// struct is initialized with `new`.
+    /// Representation of in memory struct layout.
+    /// struct DescriptorTable = [Descriptor; <queue_size>]
+    pub desc_table_ptr: *const Descriptor,
+
+    /// This is a use:errr space pointer
+    /// to the available ring, set up by the
+    /// guest. Getting access to the underling
+    /// data structure should only occur after the
+    /// struct is initialized with `new`.
+    ///
+    /// Representation of in memory struct layout.
+    /// struct AvailRing {
+    ///     flags: u16,
+    ///     idx: u16,
+    ///     ring: [u16; <queue size>],
+    ///     used_event: u16,
+    /// }
+    ///
+    /// Because all types in the AvailRing are u16,
+    /// we store pointer as *mut u16 for simplicity.
+    pub avail_ring_ptr: *mut u16,
+
+    /// This is a user space pointer
+    /// to the used ring, set up by the
+    /// guest. Getting access to the underling
+    /// data structure should only occur after the
+    /// struct is initialized with `new`.
+    ///
+    /// Representation of in memory struct layout.
+    // struct UsedRing {
+    //     flags: u16,
+    //     idx: u16,
+    //     ring: [UsedElement; <queue size>],
+    //     avail_event: u16,
+    // }
+    /// Because types in the UsedRing are different (u16 and u32)
+    /// store pointer as *mut u8.
+    pub used_ring_ptr: *mut u8,
```

Comment:
maybe

```suggestion
    pub used_ring_ptr: *mut libc::c_void,
```

?

(for all of these, really)

Reply:
Why? It is nice to know what types are behind the pointers. And it makes math a bit easier.

Reply:
well, for this specific one I had the idea because of the comment. u8 is already not the type of the stuff we point at, and `void *` is the canonical type for these opaque kinda things.

Reply:
why use type from `libc` when Rust has u8 builtin? Technically `*void` in Rust will be `*()`. But this feel very wrong. With `u8`, `u16` we at least see the underlying type. So when doing arithmetic we know that `*u8 + 1` just increases pointer by 1, while `*u16 + 1` actually moves it 2 bytes forward. With `libc::void` it is more confusing (even though technically it is same as `*u8`).